File: src/main/java/club/xiaojiawei/bean/GameRect.java
package club.xiaojiawei.bean;

import club.xiaojiawei.data.GameRationStaticData;
import club.xiaojiawei.data.ScriptStaticData;
import club.xiaojiawei.utils.GameUtil;
import club.xiaojiawei.utils.MouseUtil;
import club.xiaojiawei.utils.RandomUtil;

import java.awt.*;

/**
 * @author 肖嘉威 xjw580@qq.com
 * @date 2024/8/28 15:15
 */
public record GameRect(double left, double right, double top, double bottom) {

    public static final GameRect INVALID = new GameRect(0, 0, 0, 0);

    public Point getClickPos() {
        int realH = ScriptStaticData.GAME_RECT.bottom - ScriptStaticData.GAME_RECT.top, usableH = realH;
        int realW = ScriptStaticData.GAME_RECT.right - ScriptStaticData.GAME_RECT.left;
        int usableW = (int) (realH * GameRationStaticData.GAME_WINDOW_ASPECT_TO_HEIGHT_RATIO);
        int middleX = realW >> 1;
        int middleY = realH >> 1;
        double pointX = RandomUtil.getRandom(left, right);
        double pointY = RandomUtil.getRandom(top, bottom);
        return new Point((int) (middleX + pointX * usableW), (int) (middleY + pointY * usableH));
    }

    public void lClick() {
        GameUtil.leftButtonClick(getClickPos());
    }

    public void rClick() {
        GameUtil.rightButtonClick(getClickPos());
    }

    public void lClickMoveLClick(GameRect endRect) {
        if (endRect == null) {
            return;
        }
        Point startPos = getClickPos();
        Point endPos = endRect.getClickPos();
        GameUtil.leftButtonClick(startPos);
        GameUtil.moveMouseByLine(startPos, endPos);
        GameUtil.leftButtonClick(endPos);
    }
}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/hsscript/bean/GameRect.kt
package club.xiaojiawei.hsscript.bean

import club.xiaojiawei.hsscript.data.GAME_RECT
import club.xiaojiawei.hsscript.data.GameRationConst
import club.xiaojiawei.hsscript.utils.GameUtil
import club.xiaojiawei.util.RandomUtil
import club.xiaojiawei.hsscript.utils.SystemUtil
import java.awt.Point
import java.util.ArrayList
import java.util.function.Consumer

/**
 * @author 肖嘉威 xjw580@qq.com
 * @date 2024/8/28 15:15
 */
data class GameRect(val left: Double, val right: Double, val top: Double, val bottom: Double) {

    fun getClickPos(): Point {
        val realH: Int = GAME_RECT.bottom - GAME_RECT.top
        val usableH = realH
        val realW: Int = GAME_RECT.right - GAME_RECT.left
        val usableW = (realH * GameRationConst.GAME_WINDOW_ASPECT_TO_HEIGHT_RATIO).toInt()
        val middleX = realW shr 1
        val middleY = realH shr 1
        val pointX: Int = RandomUtil.getRandom((left * usableW).toInt(), (right * usableW).toInt())
        val pointY: Int = RandomUtil.getRandom((top * usableH).toInt(), (bottom * usableH).toInt())
        return Point(middleX + pointX, middleY + pointY)
    }

    private fun cancel() {
        GameUtil.cancelAction()
        SystemUtil.delayTiny()
    }

    fun isValid(): Boolean {
        return this != INVALID
    }

    /**
     * @param isCancel 执行前先右键取消
     */
    @JvmOverloads
    fun lClick(isCancel: Boolean = true) {
        if (isCancel) cancel()
        GameUtil.leftButtonClick(getClickPos())
    }

    fun rClick() {
        GameUtil.rightButtonClick(getClickPos())
    }

    @JvmOverloads
    fun lClickMoveLClick(endRect: GameRect?, isCancel: Boolean = true) {
        if (endRect == null) {
            return
        }
        if (isCancel) cancel()
        val startPos = getClickPos()
        val endPos = endRect.getClickPos()
        GameUtil.leftButtonClick(startPos)
        SystemUtil.delay(80, 140)
        GameUtil.moveMouse(startPos, endPos)
        SystemUtil.delay(60, 120)
        GameUtil.leftButtonClick(endPos)
    }

    fun move() {
        GameUtil.moveMouse(getClickPos())
    }

    fun move(endRect: GameRect?) {
        if (endRect == null) {
            move()
        } else {
            GameUtil.moveMouse(getClickPos(), endRect.getClickPos())
        }
    }

    fun buildAction(): GameRectAction {
        return GameRectAction(this)
    }

    class GameRectAction(private val rect: GameRect) {

        private val runnableList: MutableList<Runnable?> = ArrayList<Runnable?>()

        private var lastRect: GameRect? = null

        fun clear(): GameRectAction {
            runnableList.clear()
            lastRect = null
            return this
        }

        fun exec(): GameRectAction {
            runnableList.forEach(Consumer { obj: Runnable? -> obj!!.run() })
            return this
        }

        fun lClick(): GameRectAction {
            runnableList.add(Runnable { rect.lClick() })
            return this
        }

        fun lClick(rect: GameRect?): GameRectAction {
            runnableList.add(Runnable {
                if (rect == null) {
                    if (lastRect != null) {
                        lastRect!!.lClick()
                    }
                } else {
                    rect.lClick()
                    lastRect = rect
                }
            })
            return this
        }

        fun rClick(): GameRectAction {
            runnableList.add(Runnable { rect.rClick() })
            return this
        }

        fun rClick(rect: GameRect?): GameRectAction {
            runnableList.add(Runnable {
                if (rect == null) {
                    if (lastRect != null) {
                        lastRect!!.rClick()
                    }
                } else {
                    rect.rClick()
                    lastRect = rect
                }
            })
            return this
        }

        fun move(): GameRectAction {
            runnableList.add(Runnable { rect.move() })
            return this
        }

        fun move(endRect: GameRect?): GameRectAction {
            runnableList.add(Runnable { rect.move(endRect) })
            return this
        }
    }

    companion object {
        val INVALID: GameRect = GameRect(0.0, 0.0, 0.0, 0.0)
    }

}
